File: game/Assets/Plugins/iOS/CatPicker.swift
import Foundation
#if os(iOS)
import UIKit
import PhotosUI

// Task 50-photo/08: the two ways a photo gets in.
//
// Gallery uses PHPickerViewController, which runs OUTSIDE the app's process:
// the app receives only what the player picked and needs no photo-library
// permission at all, as long as it never asks for the PHAsset. That is one
// fewer permission prompt between a player and the only screen that matters.
// Camera uses UIImagePickerController, which does require
// NSCameraUsageDescription — there is no way around that one.
//
// The picked image comes back as a file path rather than bytes:
// UnitySendMessage carries a string, and a JPEG written to the temporary
// directory avoids inventing a second callback channel for binary data. Unity
// reads it and deletes it.
//
// Task 60-shell-build/16 VERIFY finding: OnPickFailed/OnPickUnavailable used
// to carry English sentences ("could not save the picked image: <system
// error>"), which the C# side folded straight into a player-visible message.
// That is prose crossing a native boundary into a project that keeps its
// words in one table (Shell/Copy.cs) - untranslatable by construction, and
// on a device set to another system language, `error.localizedDescription`
// could hand back text in that language, not English. Every reason below is
// now a fixed lowercase code, not a sentence; CatPicker.cs maps codes to
// copy, never displays one verbatim. NOT COMPILED as part of this change -
// verify with an iOS build.

@_silgen_name("UnitySendMessage")
private func UnitySendMessage(_ object: UnsafePointer<CChar>,
                              _ method: UnsafePointer<CChar>,
                              _ message: UnsafePointer<CChar>)

private func send(_ method: String, _ payload: String) {
    // The listener is a fixed GameObject created by the shell before any pick
    // starts; a missing one is a silent no-op on Unity's side, which is why
    // the shell also times out rather than waiting forever.
    "CatPickerListener".withCString { object in
        method.withCString { selector in
            payload.withCString { text in
                UnitySendMessage(object, selector, text)
            }
        }
    }
}

private func deliver(_ image: UIImage?) {
    guard let image = image, let jpeg = image.jpegData(compressionQuality: 0.95) else {
        send("OnPickFailed", "read_failed")
        return
    }
    let path = NSTemporaryDirectory() + "\(filePrefix)\(UUID().uuidString)\(fileSuffix)"
    do {
        try jpeg.write(to: URL(fileURLWithPath: path))
        send("OnPicked", path)
    } catch {
        // error.localizedDescription is deliberately not sent: it follows
        // the device's system language, not the game's, and would land
        // straight in a player-visible message. NSLog keeps it for whoever
        // reads device console output; it never crosses the Unity boundary.
        NSLog("[CatPicker] save_failed: \(error.localizedDescription)")
        send("OnPickFailed", "save_failed")
    }
}

private final class PickerDelegate: NSObject, PHPickerViewControllerDelegate,
                                    UIImagePickerControllerDelegate,
                                    UINavigationControllerDelegate {
    static let shared = PickerDelegate()

    func picker(_ picker: PHPickerViewController,
                didFinishPicking results: [PHPickerResult]) {
        picker.dismiss(animated: true)
        guard let provider = results.first?.itemProvider,
              provider.canLoadObject(ofClass: UIImage.self) else {
            send("OnPickCancelled", "")
            return
        }
        provider.loadObject(ofClass: UIImage.self) { object, _ in
            DispatchQueue.main.async { deliver(object as? UIImage) }
        }
    }

    func imagePickerController(_ picker: UIImagePickerController,
                               didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey: Any]) {
        picker.dismiss(animated: true)
        deliver(info[.originalImage] as? UIImage)
    }

    func imagePickerControllerDidCancel(_ picker: UIImagePickerController) {
        picker.dismiss(animated: true)
        send("OnPickCancelled", "")
    }
}

/// The prefix and suffix every file `deliver(_:)` writes, so purging only
/// ever touches this plugin's own leftovers in a directory the whole app
/// shares.
private let filePrefix = "catpick-"
private let fileSuffix = ".jpg"

/// Delete leftover picked photographs from the temporary directory.
///
/// Task 50-photo/13: the iOS counterpart of `CatPicker.purge()` /
/// `CatPickActivity.onCreate`, which empties Android's cache directory before
/// every pick "because the photograph must not outlive the run". iOS wrote to
/// `NSTemporaryDirectory()` with a fresh UUID name per pick and never revisited
/// it, so a process killed between `deliver(_:)` writing the JPEG and Unity
/// reading the path left a stranger's cat on disk indefinitely — the OS
/// reclaims `NSTemporaryDirectory()` on its own schedule, not on the next
/// launch. Called at the start of every pick, same as Android: a dead process
/// cannot clean up after itself, but the next one it starts can.
private func purge() {
    let directory = NSTemporaryDirectory()
    guard let names = try? FileManager.default.contentsOfDirectory(atPath: directory) else { return }
    for name in names where name.hasPrefix(filePrefix) && name.hasSuffix(fileSuffix) {
        try? FileManager.default.removeItem(atPath: directory + name)
    }
}

private func present(_ controller: UIViewController) {
    guard let root = UIApplication.shared.windows.first(where: { $0.isKeyWindow })?
        .rootViewController else {
        send("OnPickFailed", "no_window")
        return
    }
    root.present(controller, animated: true)
}

@_cdecl("CatPicker_openGallery")
public func CatPicker_openGallery() {
    DispatchQueue.main.async {
        purge()
        var configuration = PHPickerConfiguration()
        configuration.filter = .images
        configuration.selectionLimit = 1
        let picker = PHPickerViewController(configuration: configuration)
        picker.delegate = PickerDelegate.shared
        present(picker)
    }
}

@_cdecl("CatPicker_openCamera")
public func CatPicker_openCamera() {
    DispatchQueue.main.async {
        purge()
        guard UIImagePickerController.isSourceTypeAvailable(.camera) else {
            // The simulator has no camera, and neither do some iPads. The
            // shell shows the gallery path instead of a dead button.
            send("OnPickUnavailable", "camera")
            return
        }
        let picker = UIImagePickerController()
        picker.sourceType = .camera
        picker.delegate = PickerDelegate.shared
        present(picker)
    }
}

/// Whether a camera exists at all, so the shell can hide the button rather
/// than let the player press it and get nothing.
@_cdecl("CatPicker_hasCamera")
public func CatPicker_hasCamera() -> Bool {
    UIImagePickerController.isSourceTypeAvailable(.camera)
}
#endif

File: game/Assets/Plugins/Android/CatPicker.androidlib/src/main/java/com/catshelter/picker/CatPicker.java
package com.catshelter.picker;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.provider.MediaStore;
import android.util.Log;

import com.unity3d.player.UnityPlayer;

import java.io.File;

/**
 * Task 90-android/photo: the Android half of Shell/CatPicker.cs.
 *
 * The contract is Plugins/iOS/CatPicker.swift's, to the letter, because
 * View/CaptureScreen.cs must not know which phone it is on:
 *
 *   - openGallery() / openCamera() are fire-and-forget; the answer arrives
 *     through UnitySendMessage to the "CatPickerListener" GameObject.
 *   - Exactly one of OnPicked / OnPickCancelled / OnPickFailed /
 *     OnPickUnavailable is sent per call, so the screen never sits busy with
 *     nothing coming.
 *   - OnPicked carries a FILE PATH, not bytes. UnitySendMessage carries a
 *     string; a JPEG in the app's own cache avoids inventing a second
 *     callback channel for binary data. CatPicker.cs reads it and deletes it,
 *     exactly as it already does on iOS.
 *   - Every failure reason is a fixed lowercase code, never a sentence. That
 *     rule was bought the hard way on the iOS side (60-shell-build/16
 *     VERIFY): prose crossing a native boundary is untranslatable by
 *     construction, and an OS error string follows the device's language, not
 *     the game's. Log.w keeps the detail; only the code crosses.
 *
 * Nothing here asks for a permission, and nothing here writes outside
 * getCacheDir()/catpick. See AndroidManifest.xml for why both are true.
 */
public final class CatPicker {

    static final String TAG = "CatPicker";

    /** The GameObject Shell/CatPicker.cs creates before any pick starts. */
    static final String LISTENER = "CatPickerListener";

    /** Must match path= in res/xml/catpick_paths.xml. */
    static final String DIRECTORY = "catpick";

    /**
     * Must match android:authorities in AndroidManifest.xml, where it is
     * written as ${applicationId} + this suffix. Same fragile-but-loud
     * arrangement CatShare uses: if an applicationIdSuffix is ever added,
     * getUriForFile throws IllegalArgumentException rather than silently
     * pointing the camera somewhere else.
     */
    static final String AUTHORITY_SUFFIX = ".catpick";

    private CatPicker() {
    }

    // --- called from Shell/CatPicker.cs by name; do not rename -------------

    /**
     * Open the system photo picker. Needs no storage permission: the picker
     * runs out of process and hands over a URI already granted to us.
     */
    public static void openGallery() {
        start(CatPickActivity.MODE_GALLERY);
    }

    /**
     * Open the camera app. Needs no camera permission either, because we
     * delegate rather than open the camera ourselves - see the manifest.
     */
    public static void openCamera() {
        start(CatPickActivity.MODE_CAMERA);
    }

    /**
     * Whether taking a photograph is possible at all, so the shell can hide
     * the button rather than let the player press it and get nothing.
     *
     * Two questions, both of which have to be yes: is there a camera in this
     * device, and is there an app willing to drive it? A tablet can answer no
     * to the first; a stripped Android image can answer no to the second. The
     * resolveActivity half needs the <queries> block in the manifest to see
     * anything at all on API 30+.
     */
    public static boolean hasCamera() {
        return hasCamera(UnityPlayer.currentActivity);
    }

    /**
     * The same question asked from inside the proxy activity, where
     * UnityPlayer.currentActivity is not the context to be asking with.
     */
    static boolean hasCamera(Context context) {
        if (context == null) {
            return false;
        }
        try {
            PackageManager packages = context.getPackageManager();
            if (!packages.hasSystemFeature(PackageManager.FEATURE_CAMERA_ANY)) {
                return false;
            }
            return new Intent(MediaStore.ACTION_IMAGE_CAPTURE)
                    .resolveActivity(packages) != null;
        } catch (Exception e) {
            // A question that cannot be answered is answered "no": a hidden
            // button costs the player one alternative, a dead one costs her
            // the screen.
            Log.w(TAG, "hasCamera failed", e);
            return false;
        }
    }

    // --- internals ---------------------------------------------------------

    private static void start(final String mode) {
        final Activity activity = UnityPlayer.currentActivity;
        if (activity == null) {
            // The same code iOS sends when there is no window to present
            // from. CaptureScreen maps it to "our fault", which it is.
            Log.w(TAG, "no_window: UnityPlayer.currentActivity is null");
            send("OnPickFailed", "no_window");
            return;
        }
        // startActivity has to be called from the UI thread, and Unity's
        // scripting thread is not it.
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                try {
                    Intent proxy = new Intent(activity, CatPickActivity.class);
                    proxy.putExtra(CatPickActivity.EXTRA_MODE, mode);
                    activity.startActivity(proxy);
                } catch (Exception e) {
                    Log.w(TAG, "could not start the proxy activity", e);
                    send("OnPickFailed", "no_window");
                }
            }
        });
    }

    /** The directory every photograph passes through, created on demand. */
    static File directory(Context context) {
        File directory = new File(context.getCacheDir(), DIRECTORY);
        if (!directory.exists() && !directory.mkdirs()) {
            return null;
        }
        return directory;
    }

    /**
     * Empty the directory. Called before a pick as well as after one: the
     * photograph must not outlive the run, and a process killed between the
     * camera writing and Unity reading would otherwise leave a stranger's cat
     * on disk until the OS felt like reclaiming the cache.
     */
    static void purge(Context context) {
        try {
            File directory = new File(context.getCacheDir(), DIRECTORY);
            File[] leftovers = directory.listFiles();
            if (leftovers == null) {
                return;
            }
            for (File leftover : leftovers) {
                if (!leftover.delete()) {
                    Log.w(TAG, "could not delete " + leftover.getName());
                }
            }
        } catch (Exception e) {
            Log.w(TAG, "purge failed", e);
        }
    }

    static void send(String method, String payload) {
        try {
            UnityPlayer.UnitySendMessage(LISTENER, method, payload);
        } catch (Exception e) {
            // A missing listener is a silent no-op on Unity's side; anything
            // else here is not worth taking the app down for.
            Log.w(TAG, "UnitySendMessage(" + method + ") failed", e);
        }
    }
}
